### PR TITLE
Fix #682: Sorting the sitelist

### LINKED ISSUE #682: [Feature Request] Sorting / filtering of site list
It's practically impossible to maintain the list of inverted/ not inverted sites when the numbers grow close to or above 40 sites. I can't find any options regarding sorting or filtering these sites, which makes it really hard and cumbersome to maintain.

If what I'm saying is true, and such functionality is not already in place, I'd wish for better overveiw / filter / sorting for making maintenance of sites easier and faster. (Sort by domain, not **https//** and/or **www**.)

(Maybe some functionality to find and remove duplicates, like "https://google.com" and "google.com", would be usefull too)

### PR DESCRIPTION
**Pull request by RyanWils**
I added toggle button to sort the Sitelist, both ascending and descending. 
Steps I took to accomplish this.
image
The sortList function take in the sitelist data and a boolean to decide which way to sort. 
It first parses the url to exclude strings including (http, https and www.)
If the mode is true then it sorts ascending and if its false then it sorts descending.   

For the actual toggle button
image
I hardcoded the values and will add the local messages in for the text if everything else looks good as the local messages are stored in many different languages. 

I am pretty new to adding to definition list so, I'm unsure if I did the defining of the definition properly for the Toggle button. 
In mock.ts, definitions.d.ts and user-storage.ts
I added filterListed to the list to create this toggle button. 

This is  what it looks like on dark reader
image

### CODE REVIEW

File: src/background/user-storage.ts [hunk 1]
```diff
@@ -27,6 +27,7 @@ export default class UserStorage {
             siteList: [],
             siteListEnabled: [],
             applyToListedOnly: false,
+            filterListed: false,
```
[the remark this answers is not in the document]
Comment (RyanWils) [reply src/background/user-storage.ts]: Ill change it to something that is more descriptive

File: src/ui/popup/components/site-list-settings/index.tsx [hunk 2]
```diff
@@ -12,7 +12,22 @@ export default function SiteListSettings({data, actions, isFocused}: SiteListSet
     function isSiteUrlValid(value: string) {
         return /^([^\.\s]+?\.?)+$/.test(value);
     }
+    
+    function sortList(sitelist : string[], mode : boolean) {
```
[the remark this answers is not in the document]
Comment (RyanWils) [reply src/ui/popup/components/site-list-settings/index.tsx]: Ill change it to something that fits the information coming in and will run that npm command

File: src/ui/popup/components/site-list-settings/index.tsx [hunk 3]
```diff
@@ -12,7 +12,22 @@ export default function SiteListSettings({data, actions, isFocused}: SiteListSet
...
+            var textB = parseURL(b);
+            return mode ? textA.localeCompare(textB) : textB.localeCompare(textA)});
+        actions.changeSettings({siteList: data.settings.siteList});
+    }
 
+    function parseURL(text : string) {
+        return (text
+        .toLowerCase() // Make string lowercase
+        .replace(/^(https?):\/\//,'') // Remove http/https
+        .replace(/\www./g,'')); // Remove www.
```
[the remark this answers is not in the document]
Comment (RyanWils) [reply src/ui/popup/components/site-list-settings/index.tsx]: I will change this to detect www at the beginning

File: src/ui/popup/components/site-list-settings/index.tsx [hunk 4]
```diff
@@ -43,6 +58,15 @@ export default function SiteListSettings({data, actions, isFocused}: SiteListSet
                 )}
                 onSetShortcut={(shortcut) => actions.setShortcut('addSite', shortcut)}
             />
+            <Toggle
+                checked={data.settings.filterListed}
+                labelOn={"Ascending"}
```
[the remark this answers is not in the document]
Comment (RyanWils) [reply src/ui/popup/components/site-list-settings/index.tsx]: Ya, before I created all locales for the text I wanted to run it by you

**Comment by RyanWils** [comment]
Hi Alexander,

Thanks for review my pull request, I will make these changes.
How near is the completion of those UX improvements?

**Comment by alexanderby** [comment]
Hi @RyanWils, it's pretty far. I've finished the design itself (except the site list) and now I'm working on improving live-reloading experience (will finish tomorrow), so that the creation of the new layout will not take much time (planned for the next week). Then I will start adding some new features. All this work will be done in a separate branch, since it will be a major update.

So, if you wish, you can work on Site List sorting later (but it means that feature will not be released soon). Or you could try implementing the layout that I provided. I can help you with the layout if you wish.

Think if there is a need of storing this setting persistently, maybe it will be better to store this option in SiteListSettings state, especially for the first iteration. It will be easier to store filter query there too. See the example of working with state in Loader component.

**Comment by RyanWils** [comment]
Hi Alexander,
Thanks for the information, I will try/been trying to implement the layout you provided.
If I have any questions or need assistance I will contact you.

**Comment by alexanderby** [comment]
Great! I've just added CSS hot reloading. But there's a problem, if you close the popup, you have to reload the extension manually to see changes. I have to think how to store persistent changes, so that I could also implement popup's JS replacement without reloading the whole extension.

**Comment by RyanWils** [comment]
That's great to hear. Can't wait to see the final outcome of these changes :)

**Comment by alexanderby** [comment]
I meant it could help you working on layout, if you would merge the changes to your branch.

**Comment by RyanWils** [comment]
Sorry Alexander, I did not quite get what the hot reloading did until I thought a little because I usually develop in the loader that comes with npm run debug because I mainly worked on the JS portion of Darkreader. So every time I save it re compiles Darkreader, so i never really had an issue when implementing.

**Comment by alexanderby** [comment]
The problem is that when you make any code change, the extension is reloaded. This closes the popup, Chrome dev tools and after a minor change you have to open it again.

So, I've added CSS hot reloading, and now thinking how to refresh popover after JS changes instead of reloading the whole extension.

**Pull request closed without merging**